### PR TITLE
Issue 84

### LINKED ISSUE #84: DB handling proposition / document index
As new statements get added, proposition / document id's will change since we just increment in the transform methods. We should check if the proposition / document already exists in the database and if it does, use that ID. Otherwise, we should find the max ID and use that as the initial value rather than 1.

### PR DESCRIPTION
**Pull request by jsstevenson**
Close #84 

* For propositions and documents that don't have existing ID-like features (eg, no available PMID for a document), try to fetch their ID from the database if possible, and create a new ID if necessary.
* Add `-s <path-to-json>` option to CLI to load directly from a specified file. If you're working on adding a new source or expanding what's harvested, this will make it a little simpler to load an up-to-date set of transformed data to check against for new item IDs.

This requires being careful with your workflow. As source material and our harvest/transform methods evolve, the order in which individual items come up in the transform JSONs can change -- so, if you're running a transform, it's important to ensure the graph is loaded with the "canonical" copy of existing data.

 * Since we keep transform docs on S3, we could also have a "refresh" method to wipe the local graph and then fetch from S3 + load (???)

### CODE REVIEW

File: metakb/schemas.py [hunk 1]
```diff
@@ -91,6 +92,11 @@ class FunctionalPredicate(str, Enum):
     DOMINATE_NEGATIVE = 'causes_dominant_negative_function_of'
 
 
+Predicate = Union[PredictivePredicate, DiagnosticPredicate,
```
Comment (korikuzma) [inline metakb/schemas.py:95]: Should replace union in `Proposition` class as well with `Predicate`

**Review by korikuzma — COMMENTED** [review]
Few comments while I update my cdm files...

**Review by korikuzma — CHANGES_REQUESTED** [review]
Please fix failing tests

File: analysis/civic/examples/transform/civic_cdm_20211228.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
             "description": "A phase III clinical trial (NCT00949650) found that median progression free survival among patients with exon 19 deletions or L858R EGFR mutations (n = 308) was 13.6 months for afatinib and 6.9 months for chemotherapy (HR, 0.47; 95% CI, 0.34 to 0.65; P = 0.001).",
             "direction": "supports",
             "evidence_level": "civic.evidence_level:B",
-            "proposition": "proposition:127",
+            "proposition": "proposition:5k7StVvnNBf6qxTgXgZBCSaiyWScvS3Q",
```
[the remark this answers is not in the document]
Comment (jsstevenson) [reply analysis/civic/examples/transform/civic_cdm_20211228.json:8]: Yeah, it feels a little weird, but hashing is the only way I can think of to make IDs consistent without some sort of "seed" (like an existing, populated database) that the transform methods perform lookups against (which also feels a little weird?

**Comment by jsstevenson** [comment]
I'm going to close this for now -- I think we can make some progress given concurrent updates to the rest of MetaKB but we can return to it later

**Comment by jsstevenson** [comment]
[@korikuzma I edited the PR description as well -- may spur thoughts]

**Review by korikuzma — CHANGES_REQUESTED** [review]
Haven't looked at this much, but read updated description + CLI changes. Have a couple comments:
- Could you merge staging + resolve conflicts before I take a deeper look?
- For the new CLI option, I would maybe use a better name like `load_new_cdm` + maybe even change `load_transformed` to `load_transformed_cdms`. I think this would help make the option names clearer that the former is loading a single CDM file and the latter is loading multiple CDM files.
- For your comment about loading from s3, that issue exists [here](https://github.com/cancervariants/metakb/issues/108).

**Comment by jsstevenson** [comment]
I'm quickly realizing that a lot of the CLI options get sort of confusing between the MetaKB database and the normalizer tables

Maybe something like this (for this + #108)?

```
    @click.option(
        "--load_latest_cdms",
        "-l",
        is_flag=True,
        default=False,
        help=("Clear MetaKB database and load most recent available source CDM "
              "files. Does not run harvest and transform methods to generate "
              "new CDM files.")
    )
    @click.option(
        "--load_target_cdm",
        "-t",
        type=click.Path(exists=True, dir_okay=False, readable=True,
                        path_type=Path),
        required=False,
        help=("Load transformed CDM file at specified path. Overrides "
              "--load_normalizers_db, --force_load_normalizers_db, "
              "and --load_newest_cdms.")
    )
    @click.option(
        "--load_retrieved_cdms",
        "-r",
        default=False,
        help=("Retrieve latest CDM files from remote repository, then clear "
              "MetaKB database and load retrieved data. Overrides  "
              "--load_target_cdm, --load_normalizers_db, "
              "--force_load_normalizers_db, and --load_newest_cdms.")
    )
```

**Comment by korikuzma** [comment]
> 

Could we do `load_s3_cdms` rather than `load_retrieved_cdms`?

**Comment by jsstevenson** [comment]
Yeah, that makes sense to me

**Review by korikuzma — CHANGES_REQUESTED** [review]
Small changes. Going to re-run transforms while you address suggestions.

**Comment by korikuzma** [comment]
@jsstevenson also need to update `cli.py`. `load_transformed` --> `load_latest_cdms` and `specify_transform_file` --> `load_target_cdm`. Currently it throws an error

File: analysis/civic/examples/transform/civic_cdm_20220111.json [hunk 3]
```diff
@@ -0,0 +1,401 @@
+{
+    "statements": [
+        {
+            "id": "civic.eid:879",
```
[the remark this answers is not in the document]
Comment (jsstevenson) [reply analysis/civic/examples/transform/civic_cdm_20220111.json:4]: Ah good call -- I believe I changed the civic example script to just save to a static name but there are some old ones hanging around

**Comment by korikuzma** [comment]
@jsstevenson I'm still getting some failed tests. Some are due to me not updating my database (which I'm doing now), but I think others may not be due to this (might be wrong). Do any of your tests fail?

**Review by korikuzma — CHANGES_REQUESTED** [review]
Do you wanna go ahead and bump the variation-normalizer version in requirements? I'm getting that one pandas/sqlalchemy warning you were talking about a couple days ago

**Review by korikuzma — APPROVED** [review]

**Pull request merged**